src/mpi/nsched/mpir_nsched_progress.c

```diff
@@ -0,0 +1,157 @@
+/* -*- Mode: c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2016 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#include <assert.h>
+#include "mpiimpl.h"
+
+struct MPII_Nsched_active_comm *MPII_Nsched_active_comm_list_head = NULL;
+struct MPII_Nsched_active_comm *MPII_Nsched_active_comm_list_tail = NULL;
+
+int MPIR_Nsched_are_pending(void)
+{
+    return !!MPII_Nsched_active_comm_list_head;
+}
+
+int MPIR_Nsched_kickoff(MPIR_Nsched * s, struct MPIR_Comm *comm)
+{
+    if (comm->nsched_list_head == NULL) {
+        struct MPII_Nsched_active_comm *ac;
+
+        ac = MPL_malloc(sizeof(struct MPII_Nsched_active_comm));
+
+        ac->comm = comm;
+        ac->next = NULL;
+
+        if (MPII_Nsched_active_comm_list_head == NULL)
```

Comment:
Yes.  Also, we don't want the progress engine to spend time looking through all the existing communicators for possible NBC operations.  So we maintain a list of all the comms that have an active NBC.